src/TeamSyncer.ts

```diff
@@ -0,0 +1,317 @@
+        const {user} = (await client.users.info({ user: userId })) as UsersInfoResponse;
+        try {
+            await client.chat.postEphemeral({
+                user: channelItem.creator,
+                text: `Hint: To bridge to Matrix, run the \`/invite @${user!.name}\` command in this channel.`,
+                channel: channelItem.id,
+            });
+        } catch (ex) {
+            log.warn("Could not inform the creator to invite the bot:", ex);
+        }
+        // Create the room first.
+        try {
+            const roomId = await this.createRoomForChannel(teamId, channelItem.creator, channelItem);
+            await this.main.actionLink({
+                matrix_room_id: roomId,
+                slack_channel_id: channelItem.id,
+                team_id: teamId,
+            });
+        } catch (ex) {
+            log.error("Failed to provision new room dynamically:", ex);
+        }
+    }
+
+    public async onChannelDeleted(teamId: string, channelId: string) {
+        log.info(`${teamId} removed channel ${channelId}`);
+        if (!this.shouldTeamSync(teamId, "channel", channelId)) {
+            return;
+        }
+        const room = this.main.rooms.getBySlackChannelId(channelId);
+        if (!room) {
+            log.warn("Not unlinking channel, no room found");
+            return;
+        }
+
+        try {
+            await this.main.botIntent.sendMessage(room.MatrixRoomId, {
+                msgtype: "m.notice",
+                body: "The Slack channel bridged to this room has been deleted.",
+            });
+        } catch (ex) {
+            log.warn("Failed to send deletion notice into the room:", ex);
+        }
+
+        // Hide deleted channels in the room directory.
+        try {
+            await this.main.botIntent.getClient().setRoomDirectoryVisibility(room.MatrixRoomId, "private");
+        } catch (ex) {
+            log.warn("Failed to hide room from the room directory:", ex);
+        }
+
+        try {
+            await this.main.actionUnlink({ matrix_room_id: room.MatrixRoomId });
+        } catch (ex) {
+            log.warn("Tried to unlink room but failed:", ex);
+        }
+    }
+
+    public async onTeamMemberJoined(teamId: string, user: any) {
+        log.info(`${teamId} added user ${user}`);
+    }
+
+    public async onUserChange(teamId: string, user: any) {
+        log.info(`${teamId} added user ${user}`);
+    }
+
+    private getAliasPrefix(teamId: string) {
+        const channelConfig = this.teamConfigs[teamId].channels;
+        if (!channelConfig) {
+            return;
+        }
+        return channelConfig.alias_prefix;
+    }
+
+    private async createRoomForChannel(teamId: string, creator: string, channel: ConversationsInfo): Promise<string> {
+        let intent;
+        let creatorUserId: string|undefined;
+        try {
+            creatorUserId = (await this.main.getGhostForSlack(creator, undefined, teamId)).userId;
```

Comment:
Do we really want all these rooms to be created by a ghost and have a ghost have PL 100?